apps/emqx/src/emqx_ocsp_cache.erl

```diff
@@ -0,0 +1,521 @@
+                not_found ->
+                    ?SLOG(error, #{
+                        msg => "listener_config_missing",
+                        listener_id => ListenerID
+                    }),
+                    (ErrorResp);
+                Pattern ->
+                    Action;
+                OtherConfig ->
+                    ?SLOG(error, #{
+                        msg => "listener_config_inconsistent",
+                        listener_id => ListenerID,
+                        config => OtherConfig
+                    }),
+                    (ErrorResp)
+            end;
+        _Err ->
+            ?SLOG(error, #{
+                msg => "listener_id_not_found",
+                listener_id => ListenerID
+            }),
+            (ErrorResp)
+    end
+).
+
+%% Allow usage of OTP certificate record fields (camelCase).
+-elvis([
+    {elvis_style, atom_naming_convention, #{
+        regex => "^([a-z][a-z0-9]*_?)([a-zA-Z0-9]*_?)*$",
+        enclosed_atoms => ".*"
+    }}
+]).
+
+%%--------------------------------------------------------------------
+%% API
+%%--------------------------------------------------------------------
+
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+sni_fun(_ServerName, ListenerID) ->
+    Res =
+        try
+            fetch_response(ListenerID)
+        catch
+            _:_ -> error
+        end,
+    case Res of
+        {ok, Response} ->
+            [
+                {certificate_status, #certificate_status{
+                    status_type = ?CERTIFICATE_STATUS_TYPE_OCSP,
+                    response = Response
+                }}
+            ];
+        error ->
+            []
+    end.
+
+fetch_response(ListenerID) ->
+    case do_lookup(ListenerID) of
+        {ok, DERResponse} ->
+            {ok, DERResponse};
+        {error, invalid_listener_id} ->
+            error;
+        {error, not_cached} ->
+            ?tp(ocsp_cache_miss, #{listener_id => ListenerID}),
+            ?SLOG(debug, #{
+                msg => "fetching_new_ocsp_response",
+                listener_id => ListenerID
+            }),
+            http_fetch(ListenerID)
+    end.
+
+register_listener(ListenerID, Opts) ->
+    gen_server:call(?MODULE, {register_listener, ListenerID, Opts}, ?CALL_TIMEOUT).
+
+-spec inject_sni_fun(emqx_listeners:listener_id(), map()) -> map().
+inject_sni_fun(ListenerID, Conf0) ->
+    SNIFun = emqx_const_v1:make_sni_fun(ListenerID),
+    Conf = emqx_map_lib:deep_merge(Conf0, #{ssl_options => #{sni_fun => SNIFun}}),
+    ok = ?MODULE:register_listener(ListenerID, Conf),
+    Conf.
+
+%%--------------------------------------------------------------------
+%% gen_server behaviour
+%%--------------------------------------------------------------------
+
+init(_Args) ->
+    logger:set_process_metadata(#{domain => [emqx, ocsp, cache]}),
+    _ = ets:new(?CACHE_TAB, [
```

Comment:
Do we consider OK to lose cache on process crash?

Reply:
Excellent point :thinking: 

The effect of losing the cache would be that the response is not sent to the connecting client until it's restored.  So, if the client requires that, then it could lead to a period of clients not being able to connect (due to their own policy).

I think we could mitigate that risk by making the process supervisor the heir of this table and then making it `public`.  Does this make sense?